### PR TITLE
Support Tmux 2.4 mappings

### PR DESCRIPTION
**Pull request by thalesmello**
@bruno- 

I've finally manage to convert tmux-copycat to support both Tmux 2.4, whereas keeping Tmux 2.3 support.

Could you take a look, please?

### CODE REVIEW

**Comment by thalesmello** [comment]
Addresses #118 and #109

**Comment by bruno-** [comment]
Hey,

moving forward, do you think we need to support tmux 2.3 and below? I'd tag current `master`, drop support for 2.3 and clearly state that we're 2.4 only (maybe instructions what to do to make it work for 2.3).

Also, does this work with tmux 2.5?

**Comment by marcvs** [comment]
I was trying with tmux 2.6, no success.
prefix + c-f + entering one or two characters drops me back into my bash.

**Comment by bruno-** [comment]
From what I know support for tmux 2.4 was discontinued/dropped by maintainer after just a couple weeks because that version had some bugs. Frankly I'd be ok so skip 2.4 altogether and focus on 2.5, 2.6 and up.

**Comment by thalesmello** [comment]
@marcvs I just pushed a version that is supposed to be working. Would you be kind to test the new version, please?

@bruno- The code successfully runs with tmux 2.5 in my machine. I've taken the code that checks the version from the tmux yank project.

I think a lot of people use this plugins with old tmux versions.
Therefore, my opinion is that we should keep on supporting pre-2.4 tmux versions.

The tests are failing, but plugin appears to be working correctly. Any idea on how to debug what might be wrong with the tests, @bruno- ?

**Comment by thalesmello** [comment]
@bruno- I noticed there were version specific setup files in the tests.

Adapting those still wasn't enough to make the tests work. Could you take a look?

**Comment by marcvs** [comment]
Yep: It works. Tested the key combinations in Readme.md, 
Version tested is tmux-2.6-rc3 as shipped in debian/testing.

**Comment by thalesmello** [comment]
@bruno- I've made all the tests currently pass when running using the `./run-tests` script.

I think the next step would be to adapt the tests to make them work on travis. Any idea why they might not be working?

**Comment by thalesmello** [comment]
@bruno- I've finally made all tests pass in Travis. I don't think there is any work left in this PR to do. I'm waiting on your feedback.

**Comment by bruno-** [comment]
@thalesmello thank you very much for working on this!
I checked out your branch locally and will test it for a day. I'm on tmux 2.6.

**Comment by thalesmello** [comment]
@bruno- Any updates?

File: test/helpers/expect_helpers.exp [hunk 1]
```diff
@@ -13,6 +13,14 @@ proc exit_status_false {} {
   set exit_status 1
 }
 
+proc sync_tmux {} {
+  sleep 1.0
+}
+
+proc sync_irb {} {
+  sleep 5.0
```
Comment (bruno-) [inline test/helpers/expect_helpers.exp]: Why do we need sleep this long.

File: test/helpers/setup_tmux_conf.sh [hunk 2]
```diff
@@ -2,12 +2,50 @@
...
+		echo "run-shell '$BASE_DIR/copycat.tmux'"              >> ~/.tmux.conf
+	else
+		echo "bind-key -t vi-copy    y copy-selection"  > ~/.tmux.conf
+		echo "bind-key -t emacs-copy y copy-selection" >> ~/.tmux.conf
+		echo "run-shell '$BASE_DIR/copycat.tmux'"              >> ~/.tmux.conf
+	fi
 }
+
+# Cache the TMUX version for speed.
+tmux_version="$(tmux -V | cut -d ' ' -f 2)"
```
Comment (bruno-) [inline test/helpers/setup_tmux_conf.sh]: Don't we already have this in `scripts/helpers.sh`?

File: test/helpers/setup_tmux_conf_with_custom_searches.sh [hunk 3]
```diff
@@ -3,12 +3,51 @@
...
+		echo "bind-key -t vi-copy    y copy-selection"  > ~/.tmux.conf
+		echo "bind-key -t emacs-copy y copy-selection" >> ~/.tmux.conf
+		echo "set -g @copycat_search_C-t 'random string[[:digit:]]+'" >> ~/.tmux.conf
+		echo "run-shell '$BASE_DIR/copycat.tmux'"              >> ~/.tmux.conf
+	fi
+}
+# Cache the TMUX version for speed.
+tmux_version="$(tmux -V | cut -d ' ' -f 2)"
+
+tmux_is_at_least() {
```
Comment (bruno-) [inline test/helpers/setup_tmux_conf_with_custom_searches.sh]: 3rd time I think...

**Comment by bruno-** [comment]
@thalesmello thank you for the patience. I've been testing `copycat` for the last 10 days or so and it's been ok.

The only big thing that bothers me is support for versions below tmux 2.4. It seems to be causing bloat (and slowness) in the code and tests. Can we drop that?

**Comment by thalesmello** [comment]
@bruno- I still think it's important to maintain backwards compatibility, for those situations you want to use the plugin with an older tmux version that might already come installed with a GNU/Linux distribution.

In order to make the problem maintaining different versions less of a burden, I've pushed an update that splits the calls the different actions in a polymorphic manner, such that version specific commands become less become easier to read through. In summary, I've replaced all the if-elses by a single if-else that dynamically loads the correct file for that tmux version.

Travis is only testing Tmux 2.4, as I couldn't find an easy way to run both 2.4 and legacy tests. It's taking longer because I've increased the sleep time quite a bunch, as the I was getting intermittent tests. Increasing the sleep time solved the problem.

Take a look and tell me what you find. If you think it's still not okay to support both versions of Tmux, give me a heads up so that I can keep a local version for my personal use, and I can adapt the current pull request to support 2.4 only.

If you need contributors to the project, I'd be glad to help.

Sincerely,
Thales Mello.

**Comment by thalesmello** [comment]
@bruno- Nevermind.

I couldn't make that version work make the tests pass.

I just pushed a version with only 2.4 support.

But still, how are we going to make it convenient for people who use legacy tmux versions to use the compatible version?

TPM appears to not support branches or tags, from what I could find. Therefore, maybe we could make a github.com/tmux-plugins/tmux-copycat-legacy fork available?

**Comment by thalesmello** [comment]
@bruno- Did you have a chance to look at my new commits?

**Comment by bruno-** [comment]
@thalesmello thank you so much for working on this and having patience.

**Comment by bruno-** [comment]
Hi @thalesmello 

I've invited you as a collaborator to this project! Welcome 🎉 

Regarding the support for tmux 2.3: I added instructions for the users how they can do this [link](https://github.com/tmux-plugins/tmux-copycat#installation-for-tmux-23-and-earlier). Overall I think it's best and easiest for everyone (us and the users) if they upgrade to latest tmux version.

**Comment by john-kurkowski** [comment]
> Addresses #118 and #109

Can we close those 2 now?

**Comment by thalesmello** [comment]
Yes. Those can be closed already.

**Comment by oblitum** [comment]
It seems [this issue as mentioned at #109](https://github.com/tmux-plugins/tmux-copycat/issues/109#issuecomment-298219188) is still present in this patch, see #121.

**Pull request merged**